### PR TITLE
Templating: Support object property and array paths in interpolation expressions 

### PR DESCRIPTION
**Pull request by torkelo**
We have an immediate need to support better ways to access objects & arrays in template variable interpolations. 

See #18282

Today we have to add a bunch of seperate variables in the scopedVars props like `${__series_name}` for data links and `${__cell_0}`  `${__cell_1}` for every table column in table cell links & gauge titles. We now need to support referring to series labels as well so would have to add `__series_label_datacenter` etc for every tag to the scopedVars parameter. 

This PR tries to explore ways to support rich objects and arrays in expressions. Like `${series.name}` 
`${series.labels.datacenter}` and arrays like `${series.fields[0].name}` 

This solution is a bit hacky, tried a different approach where variable lookup always went through a lodash _.property function but for non scoped variables we also need to access the variable and not just the final value which complicated things. So for example `${variable.field}` we need to lookup the variable and evaluate the object property access. So I ended up changing the interpolation regex to capture the part after the dot as separate capture group and use that to evaluate the accessed value. So expressions like `${rows[0]}` will not be possible as the first part will need to be a variable (or scoped var) name. And then the path expression. 

TODO: 

* [ ] Explore alternative approaches
* [ ] Cache _.property function 
* [ ] Benchmark tests, any other ways we can improve performance
* [ ] To simplify the use of scopedVars that are accessed via this synax I skipped the {value, text} structure as your never interested in the text representation of a whole object in these scenarios. But it feels a bit inconsistent. Can be added back, it would not change much besides having the object wrapped in a value/text object. 
* [ ] How should we start using this, what to do with ${__series_name} and ${__cell_n} in terms of backwards compatability.

### CODE REVIEW

**Comment by torkelo** [comment]
@ryantxu @dprokop worth continuing with this approach? Or can you see other options?

**Comment by dprokop** [comment]
I like this approach @torkelo. The only thing I don't like that much is the `series.*` prefix used in the variables. Single data link relates to a series, so this looks redundant to me and I think could be solved in the variable select dropdown. But on the other hand having the `series.*` prefix makes the link more readable at first look. But I guess this is required anywayse because of z`So expressions like ${rows[0]} will not be possible as the first part will need to be a variable (or scoped var) name. And then the path expression.` ?

**Comment by ryantxu** [comment]
I like it!

We can add a scoped var for `series` - the current DataFrame - and get `${series.name}`,  `${series.labels}`, `${series.labels.xyz}`

We can also add `field` -- the current field -- and get `${field.name}`,  `${field.unit}` etc

The current `__cell_X` is a little more complicated.  With #18391 we could add the DataFrameView that gives access to the current row with either array index or field name.  Then we could have `row.time` or `row[3]`

File: public/app/features/templating/template_srv.ts [hunk 1]
```diff
@@ -225,34 +230,57 @@ export class TemplateSrv {
     return values;
   }
 
+  getFieldAccessor(fieldPath: string) {
```
Comment (dprokop) [inline public/app/features/templating/template_srv.ts]: Good idea with this simple cache! Like it

File: public/app/features/templating/template_srv.ts [hunk 2]
```diff
@@ -225,34 +230,57 @@ export class TemplateSrv {
...
+    const accessor = this.fieldAccessorCache[fieldPath];
+    if (accessor) {
+      return accessor;
+    }
+
+    return (this.fieldAccessorCache[fieldPath] = _.property(fieldPath));
```
Comment (dprokop) [inline public/app/features/templating/template_srv.ts]: Not a fan of such returns. What about changing this to:

```suggestion
   this.fieldAccessorCache[fieldPath] = _.property(fieldPath)
    return this.fieldAccessorCache[fieldPath];
```

**Review by dprokop — APPROVED** [review]
LGTM, some minor comments left

**Comment by ryantxu** [comment]
@torkelo -- Is this a duplicate that should be closed now?

**Comment by torkelo** [comment]
yea, strange. Not sure why this is open or how a dupe was made

**Pull request closed without merging**